*: add engine component

overvenus opened the pull request:
## What have you changed? (mandatory)

Tidy up rocksdb usage. #4416 reexports rocksdb at `storage::engine`, it is a good start towrads modulized tikv. This PR takes a step further, it moves all rocksdb imports to a new mod `engine` at the crate root, it helps to build a better structure, also it makes easy for introducing new engines later.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Unit tests.

## Does this PR affect documentation (docs) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

brson commented on Cargo.toml:
```diff
@@ -124,8 +124,9 @@ log_wrappers = { path = "components/log_wrappers" }
-[dependencies.rocksdb_do_not_use]
+[dependencies.engine_rocksdb]
 git = "https://github.com/pingcap/rust-rocksdb.git"
+rev = "5f49f9ef202d8ad9515de2bb05c8395c1c66b8aa"
```
Thanks for adding the rev.

zhangjinpeng1987 replied:
I think this will import some resistance for upgrade RocksDB.

brson replied:
Why?

BusyJay replied:
I think what he means is in the past, `cargo update -p rocksdb` will do all the work, but now either removing the line before update or looking up the hash manually.

I personally prefer to remove the revision too as that's the whole point of keeping a .lock file.

brson commented on components/test_raftstore/Cargo.toml:
```diff
@@ -18,3 +18,8 @@ rand = "0.3"
+[dependencies.engine_rocksdb]
+git = "https://github.com/pingcap/rust-rocksdb.git"
+rev = "5f49f9ef202d8ad9515de2bb05c8395c1c66b8aa"
+package = "rocksdb"
```
This is unfortunate, and I think unnecessary. After the previous PR all the rocksdb imports used project-wide should have been reexported through a single module in tikv, and after this pr the abstraction boundary should is `tikv::engine`. If this crate then depends concretely on rocksdb then it's violating an abstraction layer and will probably need to be removed again later. If you agree will you try to remove this and import everything needed from tikv?

overvenus replied:
You are right, I will remove it.

zhouqiang-cl commented:
/test

zhouqiang-cl commented:
/test

zhouqiang-cl commented:
/test

overvenus commented:
@brson Thank you for your review!

> It looks like there are a number of functional changes here beyond moving code into `tikv::engine`. Perhaps you could describe what their purpose is.

Actually, there is only one functional change, the error handling. This PR bring a new `engine::Error`, which is a replacement of rocksdb error, it is implemented `From<engine::Error> for raft::Error` and `From<engine::Error> for kvproto::errorpb::Error` for using `?` seamlessly in other mods.

Other changes are all about moving code with one idea in mind, `tikv::engine` should not depend on `tikv::storage`, because the storage belongs to the transaction layer above the raft layer which is in turn above the `tikv::engine` layer. The transaction layer provides a consistent transactional storage engine (rocksdb). I split a few files(mods) into different places, 
- the `rocksdb_util::event_listener` mod is split into `engine::event_listener` and `storage::engine::compaction_listener`
- split `raftstore::util` into `engine::util` and `raftstore::coprocessor`

> Is there anything written about the approach you are planning to take with the engine abstraction?

The overall goal is to provide a simple engine abstraction so we can introduce a custom engine easily. It is not about a full-featured engine, but a minimum functional engine that satisfies the raftstore mod. This PR is prior work of the goal.

zhouqiang-cl commented:
/test

zhouqiang-cl commented:
/test

BusyJay commented on components/engine/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+                .set_start_key(start_key.to_vec());
+            errorpb
+                .mut_key_not_in_region()
+                .set_end_key(end_key.to_vec());
```
Why clone it?

BusyJay commented on components/engine/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+        Ok(Some(m))
+    }
+
+    fn get_u64(&self, key: &[u8]) -> Result<Option<u64>> {
```
Is it still in use?

BusyJay commented on components/engine/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    Prefix,
+}
+
+pub struct IterOption {
```
How about just keep it as `pub struct IterOption(ReadOption)`? So that `*_bound`s doesn't have to be stored twice.

zhangjinpeng1987 replied:
Better in another pr, this pr just split modules.

zhangjinpeng1987 commented on src/raftstore/store/mod.rs:
```diff
@@ -54,8 +52,8 @@ pub use self::snap::{
 pub use self::transport::{CasualRouter, ProposalRouter, StoreRouter, Transport};
-pub use self::util::Engines;
 pub use self::worker::{KeyEntry, ReadTask, RegionTask};
+// pub use self::coprocessor::
```
Please remove this line.

siddontang commented:
Great work!!!

Connor1996 commented on src/server/debug.rs:
```diff
@@ -1318,15 +1321,11 @@ fn set_region_tombstone(db: &DB, store_id: u64, region: Region, wb: &WriteBatch)
-
-    let default_cf_size = box_try!(get_cf_size(CF_DEFAULT));
-    let write_cf_size = box_try!(get_cf_size(CF_WRITE));
+    // Empty start and end key caver all range.
```
```suggestion
    // Empty start and end key cover all range.
```

overvenus commented:
/run-integration-tests

brson commented:
Thanks for the explanation @overvenus!

BusyJay commented on benches/hierarchy/storage/mod.rs:
```diff
@@ -15,9 +15,10 @@ use criterion::{black_box, Bencher, Criterion};
 use test_util::KvGenerator;
-use tikv::storage::engine::Engine;
+use tikv::storage::kv::Engine;
 
```
Redundant blank line.

BusyJay commented on src/storage/mod.rs:
```diff
@@ -1285,14 +1267,18 @@ impl<E: Engine> Storage<E> {
             return Ok(());
         }
 
+        let cf = Self::rawkv_cf(&cf)?;
```
Is it duplicated with L1277?

BusyJay commented on src/storage/mod.rs:
```diff
@@ -945,23 +937,13 @@ impl<E: Engine> Storage<E> {
-            // operation, RocksDB will add start key to the prefix bloom, and the start key
-            // will go through function prefix_extractor. In our case the prefix_extractor
-            // is FixedSuffixSliceTransform, which will trim the timestamp at the tail. If the
-            // length of start key is less than 8, we will encounter index out of range error.
```
Why it's not a problem anymore?

overvenus replied:
Because we thought `DeleteRange` as a kv pair tiger panic when then the key length is less than 8 bytes, but actually it does not, so we can safely remove the padding.

BusyJay commented on components/engine/src/rocks/util/stats.rs:
```diff
@@ -0,0 +1,43 @@
+    for name in engine.cf_names() {
+        let handler = engine.cf_handle(name).unwrap();
+        if let Some(v) = engine.get_property_value_cf(handler, ROCKSDB_CF_STATS_KEY) {
+            writeln!(&mut s, "{}", v).unwrap();
```
Why not just `s.extend_from_slice(v)`?

overvenus commented:
/run-integration-tests

siddontang commented:
Hi @brson @BusyJay 

After this PR, can we begin to separate our code into different crates?

brson commented:
@siddontang yes. the [util](https://github.com/tikv/tikv/pull/4504) crate is the first. There are more candidates [here](https://github.com/tikv/tikv/issues/3352#issuecomment-481805380). The candidates I identified need to be done serially though. More than one person can't work on them.
